### PR TITLE
Revive Relativistic Boris Push

### PR DESCRIPTION
**Pull request by JaydenR2305**
Reimplements the relativistic Boris push integrator defined in #979 using the new `ParticleTracker` framework

Must be merged after #2704

### CODE REVIEW

**Comment by JaydenR2305** [comment]
@pheuer What are your thoughts on defining some sort of integrator class? I'm just thinking we want to make sure the integration functions all have a consistent signature so that they can be easily called from the `_push()` function.. Is there a way of doing this that doesn't require creating a new class hierarchy?

**Comment by JaydenR2305** [comment]
image
image
image
image

**Comment by JaydenR2305** [comment]
It looks like the discrepancy in the factor of `c` came from CGS units

image

**Comment by pheuer** [comment]
> It looks like the discrepancy in the factor of `c` came from CGS units

Not surprising, that happens all the time! 

These plots look great! Maybe we can have one test case as v/c=0.001 (show the pusher gives the same result for non-relativistic particles) then one at v=0.999 (show that the pusher now matches the analytical solution)?

File: src/plasmapy/simulation/particle_tracker/particle_tracker.py [hunk 1]
```diff
@@ -119,9 +136,14 @@ def __init__(
         dt=None,
         dt_range=None,
         field_weighting="volume averaged",
+        integrator="explicit_boris_relativistic",
```
Comment (pheuer) [inline src/plasmapy/simulation/particle_tracker/particle_tracker.py:139]: I think it would be better if the tracker just switched over automatically if the simulation was relativistic? We have the velocities, and there's no circumstance where someone would want to use the wrong integrator?

**Comment by JaydenR2305** [comment]
There does seem to be a discrepancy with the fitting for the classical trajectory?
image

whereas the relativistic fits quite nicely
image

**Comment by JaydenR2305** [comment]
figured it out! There was some dependence on the initial velocity. I've set the initial velocity to zero for now until we can figure out a better alternative (setting it directly to some fraction of the final drift velocity wasn't working...) This just means that we have to discard the first 20 points or so to allow the particle to reach the expected drift velocity.

**Comment by pheuer** [comment]
> figured it out! There was some dependence on the initial velocity. I've set the initial velocity to zero for now until we can figure out a better alternative (setting it directly to some fraction of the final drift velocity wasn't working...) This just means that we have to discard the first 20 points or so to allow the particle to reach the expected drift velocity.

Awesome! Kind of concerning it doesn't work with an initial velocity thought... any idea why that is? Want to make sure that's not an issue with the analytic solution or the pusher or something.

**Comment by JaydenR2305** [comment]
> Awesome! Kind of concerning it doesn't work with an initial velocity thought... any idea why that is? Want to make sure that's not an issue with the analytic solution or the pusher or something.

image
I think the derivation makes the assumption that the particle is starting from rest

**Comment by pheuer** [comment]
>I think the derivation makes the assumption that the particle is starting from rest

Ah, got it, ok that's expected then! In that case I'd say this test looks good enough?

File: src/plasmapy/diagnostics/charged_particle_radiography/synthetic_radiography.py [hunk 2]
```diff
@@ -186,6 +186,10 @@ class Tracker(ParticleTracker):
         simulation. This does not include particles that have never entered
         the grids.
 
+    relativistic_beta_threshold: `float`, optional
+        The threshold fraction of the speed of light, which once exceeded, will
+        trigger the simulation to switch to a relativistic Boris push.
```
Comment (pheuer) [inline src/plasmapy/diagnostics/charged_particle_radiography/synthetic_radiography.py:191]: Put the default value here, maybe 1%?

File: src/plasmapy/simulation/particle_integrators.py [hunk 3]
```diff
@@ -150,3 +153,71 @@ def boris_push(x, v, B, E, q, m, dt, inplace: bool = True):
...
+    m : float
+        particle mass, in SI (kg) units.
+    dt : float
+        timestep, in SI (second) units.
+
+    Notes
+    -----
+    For the basic overview of this algorithm, see `boris_push`. This
+    version, based on [1]_, applies relativistic corrections such as
+    TODO.
```
Comment (pheuer) [inline src/plasmapy/simulation/particle_integrators.py:183]: TODO

File: src/plasmapy/simulation/particle_tracker/particle_tracker.py [hunk 4]
```diff
@@ -29,6 +31,20 @@
 )
 
 
+class _IntegratorInfo(TypedDict):
```
Comment (pheuer) [inline src/plasmapy/simulation/particle_tracker/particle_tracker.py:34]: Hm, I see why you wanted to make these classes. You could do that, and just have each have an `integrate` method? Just as a way to store this information more naturally? 

Otherwise I'd say put this in the file with the push algorithm, but the class idea is better

File: src/plasmapy/simulation/particle_tracker/particle_tracker.py [hunk 5]
```diff
@@ -77,6 +93,10 @@ class to the ``run`` method.
         If specified, the calculated adaptive time step will be clamped
         between the first and second values.
 
+    relativistic_beta_threshold: `float`, optional
+        The threshold fraction of the speed of light, which once exceeded, will
+        trigger the simulation to switch to a relativistic Boris push.
```
Comment (pheuer) [inline src/plasmapy/simulation/particle_tracker/particle_tracker.py:98]: Same comment here - put default value in the docstring

File: src/plasmapy/simulation/particle_tracker/particle_tracker.py [hunk 6]
```diff
@@ -118,10 +138,18 @@ def __init__(
...
         field_weighting="volume averaged",
         req_quantities=None,
         verbose=True,
     ) -> None:
+        # By default, set the integrator to the explicit Boris push
+        # The `_push()` method may change this to a relativistic integrator if the energies of the particles
+        # exceed the threshold specified by `relativistic_beta_threshold`
+        self._integrator = _INTEGRATORS["explicit_boris"]
+        self._beta_threshold = relativistic_beta_threshold
+        self._is_relativistic = False
```
Comment (pheuer) [inline src/plasmapy/simulation/particle_tracker/particle_tracker.py:151]: If you do make the integrator a class, then `_is_relativistic can just be a property 

def _is_relativistic(self):
return self._integrator.relativistic

File: tests/simulation/test_particle_tracker.py [hunk 7]
```diff
@@ -438,3 +440,141 @@ def test_particle_tracker_stop_particles(request):
...
+        # Calculate the positions
+        # Eq. 71
+        a = γd * vd / ν
+        x = (a * γd * (ν * 𝜏 - np.sin(ν.si.value * 𝜏.si.value))).to(u.m)
+        z = (a * (np.cos(ν.si.value * 𝜏.si.value) - 1)).to(u.m)
+
+    return x, z
+
+
+def construct_field(
```
Comment (pheuer) [inline tests/simulation/test_particle_tracker.py:503]: Personal preference - if this is only used in the below test function, maybe define it as local inside that function?

File: tests/simulation/test_particle_tracker.py [hunk 8]
```diff
@@ -438,3 +440,141 @@ def test_particle_tracker_stop_particles(request):
...
+
+@pytest.mark.parametrize(
+    "regime",
+    [
+        0.01,
+        0.5,
+        0.99,
+    ],
+)
+def test_Boris_integrator_fitting(regime):
```
Comment (pheuer) [inline tests/simulation/test_particle_tracker.py:524]: This test looks great, thanks for working out those last bugs!

File: tests/simulation/test_particle_tracker.py [hunk 9]
```diff
@@ -438,3 +440,141 @@ def test_particle_tracker_stop_particles(request):
...
+    ----------
+    """
+    save_routine = IntervalSaveRoutine(0.1 * u.ns)
+
+    simulation = ParticleTracker(
+        grids=fields,
+        save_routine=save_routine,
+        termination_condition=termination_condition,
+    )
+    simulation.load_particles(
```
Comment (pheuer) [inline tests/simulation/test_particle_tracker.py:559]: Can you make a test case with an electron, too? Since that's the most likely relativistic use case

**Review by pheuer — COMMENTED** [review]
Looking good! Few minor comments before we merge

File: src/plasmapy/simulation/particle_tracker/particle_tracker.py (hunk 4 above)
Comment (JaydenR2305) [reply src/plasmapy/simulation/particle_tracker/particle_tracker.py:34]: Yeah the class idea is definitely the way to go here. Otherwise the typing gets a little dubious because Python's typing for `Callable` is very messy https://docs.python.org/3/library/typing.html#annotating-callable-objects

Comment (pheuer) [reply src/plasmapy/simulation/particle_tracker/particle_tracker.py:34]: Ok sounds great! Maybe you can prioritize these changes to this PR today and we can try to get this in today/early next week? Looks pretty close to done to me.

File: tests/simulation/test_particle_tracker.py (hunk 9 above)
Comment (JaydenR2305) [reply tests/simulation/test_particle_tracker.py:559]: Good idea! I actually caught an interesting edge case with this. In `_adaptive_timestep` we were calculating the gyroperiod using the expression:
```py
gyroperiod = 2 * np.pi * self.m / (self.q * np.max(Bmag))
```
which doesn't play well with electrons. I wrapped this in an `np.abs()` which should fix the issue

Comment (pheuer) [reply tests/simulation/test_particle_tracker.py:579]: Oops!! Yep that's a problem. To be explicit where the sign issue is, you could wrap just q in the abs, `np.abs(q)`

File: src/plasmapy/simulation/particle_integrators.py (hunk 3 above)
Comment (JaydenR2305) [reply src/plasmapy/simulation/particle_integrators.py:183]: I added some more details here. Let me know what you think! (I have next to no experience with relativity)

File: src/plasmapy/simulation/particle_tracker/particle_tracker.py (hunk 4 above)
Comment (JaydenR2305) [reply src/plasmapy/simulation/particle_tracker/particle_tracker.py]: So I have it implemented but I'm not really happy with how I handle `is_relativistic`: https://github.com/PlasmaPy/PlasmaPy/blob/c6aee176c601687992d25e50684ad2936ff33a82/src/plasmapy/simulation/particle_integrators.py#L24-L28

I need a clean way of indicating that this property has to be defined by specifying it in the abstract class somewhere while also not requiring the class to be instantiated-- something like an "abstract property"

Comment (pheuer) [reply src/plasmapy/simulation/particle_tracker/particle_tracker.py]: Does abstract property from ABC not do what you want? https://docs.python.org/3/library/abc.html#abc.abstractproperty

File: src/plasmapy/simulation/particle_integrators.py (hunk 3 above)
Comment (pheuer) [reply src/plasmapy/simulation/particle_integrators.py]: Looks good to me!

File: src/plasmapy/simulation/particle_tracker/particle_tracker.py (hunk 4 above)
Comment (JaydenR2305) [reply src/plasmapy/simulation/particle_tracker/particle_tracker.py]: It does, but part of the issue is that we shouldn't be requiring the user to instantiate the integrator (which is required with abstract properties.) I might just stick with the `TypedDict` approach and expose the `AbstractIntegrator` as such a dict

Comment (pheuer) [reply src/plasmapy/simulation/particle_tracker/particle_tracker.py]: I had assumed the classes would all be defined here, then in tracker the function would take a keyword argument to specify which tracker to use, then instantiate it internally, `self._integrator = BorisIntegrator()` then use that for the push function? Maybe I don't understand the issue here?

The documentation is a good thing to think of: at least in this module, we want documentation for each integrator method which the classes seem a natural match for.

Comment (JaydenR2305) [reply src/plasmapy/simulation/particle_tracker/particle_tracker.py]: I'll switch it over to instantiate the integrator, definitely makes it easier to ensure the relevant documentation is included

File: src/plasmapy/simulation/particle_tracker/particle_tracker.py (hunk 1 above)
Comment (JaydenR2305) [reply src/plasmapy/simulation/particle_tracker/particle_tracker.py]: According to the profiler library they both have a percall time of 0.000 so I think it's safe to say they're close enough? What do you think

File: tests/simulation/test_particle_tracker.py (hunk 7 above)
Comment (JaydenR2305) [reply tests/simulation/test_particle_tracker.py]: What are your thoughts on moving all of the relevant functions into a class? https://github.com/PlasmaPy/PlasmaPy/blob/2f25b6c847caa2d7a69d815af522f116d8c8772b/tests/simulation/test_particle_tracker.py#L445-L662

Comment (pheuer) [reply tests/simulation/test_particle_tracker.py]: Looks like a good idea to me!

**Review by pheuer — APPROVED** [review]
This looks great, thanks for polishing off this neglected old PR!!
